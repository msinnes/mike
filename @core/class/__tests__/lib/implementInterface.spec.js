const classLoader = require('../../src/loaders/class');

const implementInterface = require('../../src/lib/implementInterface');

const iAdder = require('../../examples/interfaces/iAdder');
const iNamed = require('../../examples/interfaces/iNamed');

const iNamedAdder = iAdder.extend(iNamed);

function NamedAdder() {
  this.name = 'NamedAdder';
  this.value = 1;
  this.addOne = function() {};
  this.addTwo = function() {};
}

function UnnamedAdder() {
  this.value = 1;
  this.addOne = function() {};
  this.addTwo = function() {};
}

const LoadedNamedAdder = classLoader(NamedAdder);
const ImplementsINamedAdder = implementInterface(LoadedNamedAdder, iNamedAdder);
const LoadedUnnamedAdder = classLoader(UnnamedAdder);
const ImplementsIAdder = implementInterface(LoadedUnnamedAdder, iAdder);
const FailsImplementsINamedAdder = implementInterface(LoadedUnnamedAdder, iNamedAdder);

describe('implementInterface', () => {
  it('should be a function', () => {
    expect(implementInterface).toBeDefined();
    expect(implementInterface).toBeInstanceOf(Function);
  });

  it('should implement an interface', () => {
    expect(() => {
      new ImplementsIAdder();
    }).not.toThrow();
    expect(() => {
      new ImplementsINamedAdder();
    }).not.toThrow();

    expect(() => {
      new FailsImplementsINamedAdder();
    }).toThrowErrorMatchingSnapshot();

    expect(ImplementsIAdder.Class._implements).toEqual(iAdder);
    expect(ImplementsINamedAdder.Class._implements).toEqual(iNamedAdder);
    expect(FailsImplementsINamedAdder.Class._implements).toEqual(iNamedAdder);
  });
});