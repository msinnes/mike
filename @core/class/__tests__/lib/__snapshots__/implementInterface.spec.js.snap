// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`implementInterface should implement an interface 1`] = `
"ClassError: Runtime Interface Error
name: Expected StringType"
`;
